common/fs/src/cache/tailed_file.rs

```diff
@@ -107,39 +120,341 @@ impl Stream for Lines {
             }
         }
 
-        reader.offset += n;
+        *offset += n;
 
         Metrics::fs().add_bytes(n);
         Poll::Ready(Some(Ok(s)))
     }
 }
 
+pub struct LazyLines {
+    reader: Arc<Mutex<TailedFileInner>>,
+    current: Option<Arc<Mutex<TailedFileInner>>>,
+    read: usize,
+    path: usize,
+    paths: Vec<String>,
+}
+
+impl LazyLines {
+    pub fn new(reader: Arc<Mutex<TailedFileInner>>, paths: Vec<String>) -> Self {
+        Self {
+            reader,
+            current: None,
+            read: 0,
+            path: 0,
+            paths,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LazyLineSerializer {
+    annotations: Option<KeyValueMap>,
+    app: Option<String>,
+    env: Option<String>,
+    host: Option<String>,
+    labels: Option<KeyValueMap>,
+    level: Option<String>,
+    meta: Option<Value>,
+
+    path: String,
+
+    reader: Arc<Mutex<TailedFileInner>>,
+}
+
+#[async_trait]
+impl IngestLineSerialize<String, bytes::Bytes, std::collections::HashMap<String, String>>
```

Comment:
I think there is a refactoring possible to extract the has_* methods out as defaults using the LineMeta trait. It will require a change in the client too